chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,575 @@
+}
+
+/// The reward is distributed among delegations proportionally to their balance
+fn distribute_delegations_pos_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_delegations_reward: Amount,
+) -> Result<Option<Vec<PoSAccountingUndo>>, ConnectTransactionError> {
+    accounting_adapter
+        .accounting_delta()
+        .get_pool_delegations_shares(pool_id)?
+        .map(|delegation_shares| {
+            let total_delegations_balance =
+                delegation_shares.values().try_fold(Amount::ZERO, |acc, v| {
+                    (acc + *v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                        block_id, pool_id,
+                    ))
+                })?;
+
+            if total_delegations_balance != Amount::ZERO {
+                let rewards_per_delegation = calculate_rewards_per_delegation(
+                    &delegation_shares,
+                    block_id,
+                    total_delegations_balance,
+                    total_delegations_reward,
+                )?;
+
+                let distribute_remainder_undo = distribute_delegations_reward_remainder(
+                    accounting_adapter,
+                    block_id,
+                    pool_id,
+                    total_delegations_reward,
+                    &rewards_per_delegation,
+                )?;
+
+                let delegation_undos = rewards_per_delegation
+                    .iter()
+                    .map(|(delegation_id, reward)| {
+                        accounting_adapter
+                            .operations(TransactionSource::Chain(block_id))
+                            .delegate_staking(*delegation_id, Amount::from_atoms(*reward))
+                            .map_err(ConnectTransactionError::PoSAccountingError)
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                Ok(distribute_remainder_undo
+                    .unwrap_or_default()
+                    .into_iter()
+                    .chain(delegation_undos.into_iter())
+                    .collect())
+            } else {
+                // if total balance of all delegations is 0 then give the reward to the pool's owner
+                let increase_pool_balance_undo = accounting_adapter
+                    .operations(TransactionSource::Chain(block_id))
+                    .increase_pool_owner_balance(pool_id, total_delegations_reward)?;
+                Ok(vec![increase_pool_balance_undo])
+            }
+        })
+        .transpose()
+}
+
+fn calculate_rewards_per_delegation(
+    delegation_shares: &BTreeMap<DelegationId, Amount>,
+    block_id: Id<Block>,
+    total_delegations_balance: Amount,
+    total_delegations_reward: Amount,
+) -> Result<Vec<(DelegationId, u128)>, ConnectTransactionError> {
+    let total_delegations_balance = Uint256::from_amount(total_delegations_balance);
```

Comment:
Since we're doing the arithmetic here, let's make it more robust. Please add a static_assertion from that station_assertions crate where we assert that the Amount size is 16 bytes. This way, if that ever changes in the future, we know that Uint256 should be updated. You can use `static_assertions::assert_eq_size()`